contributors/design-proposals/api-machinery/customresource-conversion-webhook.md

```diff
@@ -128,18 +128,37 @@ type CustomResourceConversionWebhook {
+
+Hence the defaulting either implicit or explicit has the potential to break backward compatibility as we have two sets of fields representing the same feature. Refer to [this document](http://bit.ly/k8s-crd-per-version-defaulting) for more details.
+
+Hence the solution is the two sets of fields cannot be set at the same time. These validation rules need to be applied:
+
+* Either top level X or per version X can be set not both. This rule applies to individual X’s not the whole set. E.g. top level Schema can be set while per version subresources are set.
+* Per version X cannot be the same. E.g. if all per version schema are the same, the CRD object will be rejected with an error message asking the user to use the top level schema.
+
+The second validation garantees a clean moving to CRD-v1 (apiextensions.k8s.io/v1). The desired state of CRD-v1 is to only have per-version X and get rid of the top level X. To achieve that, these are conversion rules (and again these rules apply to individual fields not the whole set):
+
+* v1beta1->v1: If top level X is set in v1beta1, then it will be copied to all versions in v1.
+* v1beta1->v1: If per version X are set in v1beta1, then they will be used for per version X in v1.
+* v1->v1beta1: If all per-version X are the same in v1, they will be copied to top level X in v1beta1
+* v1->v1beta1: Otherwise, they will be used as per version X in v1beta1
+
+There is one corner case that if the user of v1beta1 set all per version X the same, then the conversion above is not round trippable. That's the case the above validation protects from.
 
 ### Support Level
 
 The feature will be alpha in the first implementation and will have a feature gate that is defaulted to false. The roll-back story with a feature gate is much more clear. if we have the features as alpha in kubernetes release Y (>X where the feature is missing) and we make it beta in kubernetes release Z, it is not safe to use the feature and downgrade from Y to X but the feature is alpha in Y which is fine. It is safe to downgrade from Z to Y (given that we enable the feature gate in Y) and that is desirable as the feature is beta in Z.
-
+On downgrading from a Z to Y, stored CRDs can have per version fields set. While the feature gate can be off on Y (alpha cluster), it is dangerous to disable per-version Schema Validation or Status subresources as it makes the status field mutable and validation on CRs will be disabled. Thus the feature gate in Y only protects adding per version fields not the actual behaviour. This way even if we downgrade from Z to Y, per version validations and subresources will be honored. This will not be the case for webhook conversion itself. The feature gate will also protect the implementation of webhook conversion and alpha cluster with disabled feature gate will return error for CRDs with webhook conversion (that are created with a future version of the cluster).
```

Comment:
is this cover what you are asking? ... `Thus the feature gate in Y only protects adding per version fields not the actual behaviour.` ....

Reply:
I was concerned if it's clear that we allows updating per version X when gate is off, if the CRD object already uses the per version X; also different from the other validation rules, I suppose this feature gate validation rule applies to the whole set instead of individual X’s? 

E.g. when gate is off, for a CRD that already uses per version Schema, users can update the CRD to use per version Subresources from using top-level Subresources previously.

Maybe it's just me :) I'm willing to LGTM after the typo reported by CI is fixed.

Reply:
Added more detail, does it look good now?